src/main/java/slimeknights/tconstruct/library/tools/ToolCore.java

```diff
@@ -142,290 +199,576 @@ public int getHarvestLevel(ItemStack stack, ToolType toolClass, @Nullable Player
+        this.breakExtraBlock(itemstack, player.getEntityWorld(), player, extraPos, pos);
+      }
+    }
 
+    /*// this is a really dumb hack.
+    // Basically when something with silktouch harvests a block from the offhand
+    // the game can't detect that. so we have to switch around the items in the hands for the break call
+    // it's switched back in onBlockDestroyed
+    if (DualToolHarvestUtil.shouldUseOffhand(player, pos, player.getHeldItemMainhand())) {
+      ItemStack off = player.getHeldItemOffhand();
+
+      this.switchItemsInHands(player);
+      // remember, off is in the mainhand now
+      CompoundNBT tag = off.getOrCreateTag();
+      tag.putLong(TAG_SWITCHED_HAND_HAX, player.getEntityWorld().getGameTime());
+      off.setTag(tag);
+    }*/
+
+    return this.breakBlock(itemstack, pos, player);
+  }
 
-  /* Information */
+  /**
+   * Called to break the base block, return false to perform no breaking
+   * @param itemstack Tool ItemStack
+   * @param pos       Current position
+   * @param player    Player instance
+   * @return true if the normal block break code should be skipped
+   */
+  // todo: find a better way to solve this and breakExtraBlock?
+  protected boolean breakBlock(ItemStack itemstack, BlockPos pos, PlayerEntity player) {
+    return super.onBlockStartBreak(itemstack, pos, player);
+  }
+
+  /**
+   * Called when an AOE block is broken by the tool. Use to override the block breaking logic
+   *
+   * @param tool   Tool ItemStack
+   * @param world  World instance
+   * @param player Player instance
+   * @param pos    Current position
+   * @param refPos Base position
+   */
+  protected void breakExtraBlock(ItemStack tool, World world, PlayerEntity player, BlockPos pos, BlockPos refPos) {
+    AoeToolInteractionUtil.breakExtraBlock(tool, world, player, pos, refPos);
+  }
+
+  /* Trait interactions */
+
+  @Override
+  public void inventoryTick(ItemStack stack, World worldIn, Entity entityIn, int itemSlot, boolean isSelected) {
+    super.inventoryTick(stack, worldIn, entityIn, itemSlot, isSelected);
+
+    this.onUpdateTraits(stack, worldIn, entityIn, itemSlot, isSelected);
+  }
 
+  protected void onUpdateTraits(ItemStack stack, World worldIn, Entity entityIn, int itemSlot, boolean isSelected) {
+    final boolean isSelectedOrOffhand = isSelected || (entityIn instanceof PlayerEntity && ((PlayerEntity) entityIn).getHeldItemOffhand() == stack);
+
+    TraitUtil.forEachTrait(stack, trait -> trait.onUpdate(stack, worldIn, entityIn, itemSlot, isSelectedOrOffhand));
+  }
+
+  /* Information */
 
   @Override
+  @OnlyIn(Dist.CLIENT)
   public void addInformation(ItemStack stack, @Nullable World worldIn, List<ITextComponent> tooltip, ITooltipFlag flagIn) {
-    // todo: evaluate if we want to keep the material info
+    boolean shift = Util.isShiftKeyDown();
+    boolean ctrl = Util.isCtrlKeyDown();
```

Comment:
ITooltipFlag do anything for us? Or is that advanced tooltips?

Reply:
ITooltipFlag is for advanced tooltips